File: src/type_builder.js
//Provides: caml_type_builder_create
//Requires: Binaryen
function caml_type_builder_create(size) {
  return new Binaryen.TypeBuilder(size);
}

//Provides: caml_type_builder_grow
function caml_type_builder_grow(builder, count) {
  builder.grow(count);
}

//Provides: caml_type_builder_get_size
function caml_type_builder_get_size(builder) {
  return builder.getSize();
}

//Provides: caml_type_builder_set_signature_type
function caml_type_builder_set_signature_type(
  builder,
  index,
  paramTypes,
  resultTypes
) {
  builder.setSignatureType(index, paramTypes, resultTypes);
}

//Provides: caml_type_builder_set_struct_type
//Requires: caml_list_to_js_array
//Requires: caml_js_from_bool
function caml_type_builder_set_struct_type(
  builder,
  index,
  fieldTypes,
  fieldPackedTypes,
  fieldMutables,
  numFields
) {
  var types = caml_list_to_js_array(fieldTypes);
  var packedTypes = caml_list_to_js_array(fieldPackedTypes);
  var mutables = caml_list_to_js_array(fieldMutables);
  var fields = types.map(function (type, idx) {
    return {
      type,
      packedType: packedTypes[idx],
      mutable: caml_js_from_bool(mutables[idx]),
    };
  });
  builder.setStructType(index, fields);
}
//Provides: caml_type_builder_set_struct_type__bytecode
//Requires: caml_type_builder_set_struct_type
function caml_type_builder_set_struct_type__bytecode() {
  return caml_type_builder_set_struct_type(
    arguments[0],
    arguments[1],
    arguments[2],
    arguments[3],
    arguments[4],
    arguments[5]
  );
}

//Provides: caml_type_builder_set_array_type
function caml_type_builder_set_array_type(
  builder,
  index,
  elementType,
  elementPackedType,
  elementMutable
) {
  builder.setArrayType(index, elementType, elementPackedType, elementMutable);
}

//Provides: caml_type_builder_get_temp_heap_type
function caml_type_builder_get_temp_heap_type(builder, index) {
  return builder.getTempHeapType(index);
}

//Provides: caml_type_builder_get_temp_tuple_type
//Requires: caml_list_to_js_array
function caml_type_builder_get_temp_tuple_type(builder, types, numTypes) {
  var tupleTypes = caml_list_to_js_array(types);
  return builder.getTempTupleType(tupleTypes);
}

//Provides: caml_type_builder_get_temp_ref_type
//Requires: caml_js_from_bool
function caml_type_builder_get_temp_ref_type(builder, heapType, nullable) {
  return builder.getTempRefType(heapType, caml_js_from_bool(nullable));
}

//Provides: caml_type_builder_set_sub_type
function caml_type_builder_set_sub_type(builder, index, superType) {
  builder.setSubType(index, superType);
}

//Provides: caml_type_builder_set_open
function caml_type_builder_set_open(builder, index) {
  builder.setOpen(index);
}

//Provides: caml_type_builder_create_rec_group
function caml_type_builder_create_rec_group(builder, index, length) {
  builder.createRecGroup(index, length);
}

//Provides: caml_type_builder_build_and_dispose
//Requires: caml_js_to_array
function caml_type_builder_build_and_dispose(builder) {
  try {
    var types = builder.buildAndDispose();
    return [0, caml_js_to_array(types)];
  } catch (e) {
    return [1, [0, e.index, e.reason]];
  }
}

File: src/type_builder.ml
type t

type struct_field = {
  type_ : Type.t;
  packed_type : Packed_type.t;
  mutable_ : bool;
}

type error =
  | SelfSupertype of int
  | InvalidSupertype of int
  | ForwardSupertypeReference of int
  | ForwardChildReference of int

external make : int -> t = "caml_type_builder_create"
external grow : t -> int -> unit = "caml_type_builder_grow"
external size : t -> int = "caml_type_builder_grow"

external set_signature_type : t -> int -> Type.t -> Type.t -> unit
  = "caml_type_builder_set_signature_type"

external set_struct_type :
  t -> int -> Type.t list -> Packed_type.t list -> bool list -> int -> unit
  = "caml_type_builder_set_struct_type__bytecode" "caml_type_builder_set_struct_type"

let set_struct_type builder index fields =
  let split_fields fields =
    List.fold_right
      (fun { type_; packed_type; mutable_ } (types, packed_types, mutables) ->
        (type_ :: types, packed_type :: packed_types, mutable_ :: mutables))
      fields ([], [], [])
  in
  let types, packed_types, mutables = split_fields fields in
  set_struct_type builder index types packed_types mutables (List.length types)

external set_array_type : t -> int -> Type.t -> Packed_type.t -> bool -> unit
  = "caml_type_builder_set_array_type"

external get_temp_heap_type : t -> int -> Heap_type.t
  = "caml_type_builder_get_temp_heap_type"

external get_temp_tuple_type : t -> Type.t list -> Type.t
  = "caml_type_builder_get_temp_tuple_type"

external get_temp_ref_type : t -> Heap_type.t -> bool -> Type.t
  = "caml_type_builder_get_temp_ref_type"

external set_sub_type : t -> int -> Type.t -> unit
  = "caml_type_builder_set_sub_type"

external set_open : t -> int -> unit = "caml_type_builder_set_open"

external create_rec_group : t -> int -> int -> unit
  = "caml_type_builder_create_rec_group"

external build_and_dispose : t -> (Type.t array, int * int) result
  = "caml_type_builder_build_and_dispose"

let build_and_dispose builder =
  match build_and_dispose builder with
  | Ok types -> Ok (Array.to_list types)
  | Error (idx, 0) -> Error (SelfSupertype idx)
  | Error (idx, 1) -> Error (InvalidSupertype idx)
  | Error (idx, 2) -> Error (ForwardSupertypeReference idx)
  | Error (idx, _) -> Error (ForwardChildReference idx)
